README.md
[![R-CMD-check](https://github.com/erocoar/gghalves/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/erocoar/gghalves/actions/workflows/R-CMD-check.yaml)
[![CRAN_Release_Badge](http://www.r-pkg.org/badges/version-ago/gghalves)](https://CRAN.R-project.org/package=gghalves)
[![CRAN_Download_Badge](http://cranlogs.r-pkg.org/badges/grand-total/gghalves)](https://CRAN.R-project.org/package=gghalves)

### About
`gghalves` makes it easy to compose your own `half-half` plots via `ggplot2`. Think displaying a boxplot next to jittered points, or violin plots side by side with dotplots. 

### Installation
`gghalves` can be installed via GitHub:

```r
if (!require(devtools)) {
    install.packages('devtools')
}
devtools::install_github('erocoar/gghalves')
```

### Features
`gghalves` adds `_half_` extensions to selected `geom`s:
- `geom_half_boxplot`
- `geom_half_violin`
- `geom_half_point`

All of them have a `side` argument that can be either `l` for left or `r` for right half. As a special case there is also `geom_half_dotplot` to support half dotplots that dodge aesthetics such as `fill` and `colour`. 

`gghalves` also works well with other `ggplot2` geoms and extensions such as `geom_dotplot` and `ggbeeswarm::geom_beeswarm`.

![Features](https://i.imgur.com/7APxx1t.png)

### Roadmap
Generally, I hope to add more features over time. If you would like to see a certain feature, please open an issue.


NEWS.md
# gghalves 0.1.4
## Bug Fixes
- fixed compatibility of `geom_half_boxplot()` with `ggplot2` version 3.4.

# gghalves 0.1.3
## Bug Fixes
- fixed `geom_half_violin()` bug where panel plots (without any grouping aesthetic) would no longer work

# gghalves 0.1.2
## New Features
- `geom_half_violin()` now accepts an optional `split` aesthetic to compare a binary variable side-by-side (see vignette)
- `geom_half_violin()` now accepts a vector for the `side` argument, supporting different sizes per plotting group

## Bug Fixes
- fixed missing import of `remove_missing` function
- fixed missing import of `translate_shape_string` function


# gghalves 0.1.1
## Minor Changes
- `transformation_params` argument of `geom_half_point` has been removed (after previously having been deprecated). Instead, pass `position_*(params)` with the `transformation` argument instead.
- `range_scale` has been set to a default value of `0.75` for `geom_half_point_panel` to match `geom_half_point`.
## Bug Fixes
- fixed a bug in `geom_half_point` where jitter width defaulted to the default value of `ggplot2::PositionJitter` rather than a custom calculation. `range_scale` should work better as a result, and the geom can be used with small values without problems.

# gghalves 0.1.0

## New features
- `geom_half_point_panel` for jittering points but allowing grouped aesthetics inside that jitter (rather than dodging per group). 

## Major Changes
- `transformation_params` argument of `geom_half_point` has been deprecated in favor of passing the arguments directly to `transformation` by evaluating `position_*(arguments)` functions that return `Position` objects.

## Minor changes
- `geom_half_point` now supports passing of positions other than `PositionJitter` and `PositionIdentity` to `transformation` -- given they have a `compute_panel` function.
- added `range_scale` parameter to `geom_half_point`. When no `width` argument is passed in `transformation_params`, `range_scale` defaults to `0.5`, which leads to the points using half of their allotted space. `1` would have them end up using all of their allotted space. 
- added default `width` and `height` values when `width` or `height` are included names in the `transformation_params` list but set to `NULL`. This leads to sensible defaults for `Position`s that are not `PositionJitter`. 


## Bug Fixes
- removed range scaling of `geom_half_point` after points were already jittered
- fixed y-axis limits being set to # of factor levels when the maximum y-value is lower than the # of factor values
- accurate naming of `geom_half_violin` and `geom_half_point` grobs by drawing internally rather than calling `Geom*$draw_panel`

# gghalves 0.0.1

## Major changes
- First commit

## New features
- `geom_half_boxplot` for half-boxplots.
- `geom_half_dotplot` for aligned dotplots that dodge fill and color.
- `geom_half_violin` for half-violin plots.
- `geom_half_point` for half-jitterplots.

R/geom_half_point.R
#' Points with jitter for half geoms.
#'
#' @inheritParams ggplot2::geom_point
#' @param side The side on which to draw the half violin plot. "l" for left, "r" for right, defaults to "l".
#' @param transformation An evaluated `position_*()` function yielding a `Position` object with specified parameters to calculate the transformation of the points. Defaults to `ggplot2::position_jitter()`.
#' @param range_scale If no `width` argument is specified in `transformation`, `range_scale` is used to determine the width of the jitter. Defaults to `0.75`, which is half of the allotted space for the jitter-points, whereas `1` would use all of the allotted space.
#' @importFrom ggplot2 layer position_jitter .pt .stroke
#' @importFrom grid pointsGrob gpar
#' @examples 
#' ggplot(iris, aes(x = Species, y = Petal.Width, fill = Species)) + 
#'   geom_half_point()
#'   
#' ggplot(iris, aes(x = Species, y = Petal.Width, fill = Species)) + 
#'   geom_half_point(side = "l")
#' @export
geom_half_point <- function(
  mapping = NULL, data = NULL,
  stat = "HalfPoint", position = "dodge2",
  ...,
  side = "r",
  transformation = position_jitter(),
  # transformation_params = list(width = NULL, height = NULL),
  range_scale = .75,
  na.rm = FALSE,
  show.legend = NA,
  inherit.aes = TRUE) {
    layer(
      data = data,
      mapping = mapping,
      stat = stat,
      geom = GeomHalfPoint,
      position = position,
      show.legend = show.legend,
      inherit.aes = inherit.aes,
      params = list(
        side = side,
        transformation = transformation,
        # transformation_params = transformation_params,
        range_scale = range_scale,
        na.rm = na.rm,
        ...
      )
    )
}

#' @rdname gghalves-extensions
#' @format NULL
#' @usage NULL
#' @importFrom ggplot2 ggproto Geom GeomBoxplot GeomPoint alpha .pt .stroke
#' @importFrom grid pointsGrob gpar
#' @export
GeomHalfPoint <- ggproto(
  "GeomHalfPoint", 
  Geom,
  required_aes = c("x", "y"),
  non_missing_aes = c("size", "shape", "colour"),
  default_aes = aes(
    shape = 19, 
    colour = "black", 
    size = 1.5, 
    fill = NA,
    alpha = NA, 
    stroke = 0.5
    ),
  
  setup_data = function(data, params) {
    x_data    <- GeomBoxplot$setup_data(data, NULL)
    data$xmin <- x_data$xmin
    data$xmax <- x_data$xmax
    data
  },

  draw_group = function(
    data, panel_params, coord, na.rm = FALSE, side = "r", 
    transformation = position_jitter(),
    # transformation_params = list(width = NULL, height = NULL),
    range_scale = .75) {

    # if (isFALSE(isTRUE(all.equal(transformation_params, list(width = NULL, height = NULL))))) {
    #   warning("Argument deprecated.\n Use `transformation = position_*(params)` instead of passing the params via `transformation_params`")
    # }
    
    if (is.character(data$shape)) {
      data$shape <- ggplot2:::translate_shape_string(data$shape)
    }
    
    xrange <- data$xmax - data$xmin
    x_add  <- (xrange / 4) * switch((side == "r") + 1, -1, 1)
    data$x <- data$x + x_add

    # Add Position Transformation
    transformation_params_new <- transformation$setup_params(data)
    transformation_df <- data_frame0(
      x     = data$x,
      y     = data$point_y[[1]],
      PANEL = 1,
      group = -1
    )
    
    
    null_width <- FALSE
    name_width <- "width" %in% names(transformation)
    if (isTRUE(name_width)) {
      null_width <- is.null(transformation[["width"]])
    }
    if ((!name_width || null_width) && "width" %in% names(transformation_params_new)) {
      transformation_params_new$width <- xrange / (4 / range_scale)
    }
    
    null_height <- FALSE
    name_height <- "height" %in% names(transformation)
    if (isTRUE(name_height)) {
      null_height <- is.null(transformation[["height"]])
    }
    if ((!name_height || null_height) && "height" %in% names(transformation_params_new)) {
      transformation_params_new$height <- ggplot2::resolution(data$point_y[[1]], zero = FALSE) * 0.4
    }
    
    # Add deprecated transformation_params args to new transformation_params list
    # par_idx <- !sapply(transformation_params, is.null)
    # transformation_params_new[names(transformation_params)[par_idx]] <- transformation_params[par_idx]
    # transformation_params <- transformation_params_new
    # 
    # # deprecated
    # if ("width" %in% names(transformation_params)) {
    #   transformation_params$width <- transformation_params$width %||% xrange / (4 / range_scale)
    # }
    # # deprecated
    # if ("height" %in% names(transformation_params)) {
    #   transformation_params$height <- transformation_params$height %||% 
    #     ggplot2::resolution(data$point_y[[1]], zero = FALSE) * 0.4
    # }

    # if (is(transformation, "PositionJitter")) {
    #   transformation_params$width  <- transformation_params$width %||% xrange / (4 / range_scale)
    #   transformation_params$height <- transformation_params$height %||% 
    #     ggplot2::resolution(data$point_y[[1]], zero = FALSE) * 0.4
    # }

    if (is(transformation, "PositionIdentity") || is(transformation, "PositionJitter")) {
      trans_positions <- transformation$compute_layer(
        transformation_df,
        transformation_params_new
      )
    } else {
      trans_positions <- transformation$compute_panel(
        transformation_df,
        transformation_params_new
      )
    }

    # if (length(unique(trans_positions$x)) > 1L) {
    #   if (side == "r") {
    #     trans_positions$x <- (data$xmax - data$x) * (
    #       trans_positions$x - min(trans_positions$x)) / (
    #         max(trans_positions$x) - min(trans_positions$x)) + data$x #- 0.045
    #   } else {
    #     trans_positions$x <- (data$x - data$xmin) * (
    #       trans_positions$x - min(trans_positions$x)) / (
    #         max(trans_positions$x) - min(trans_positions$x)) + data$xmin #+ 0.045
    #   }
    # } #TODO parameterize left/right-shift
      

    point_df <- data_frame0(
      colour = data$colour,
      shape  = data$shape,
      x      = trans_positions$x,
      y      = trans_positions$y,
      PANEL  = data$PANEL,
      group  = data$group,
      size   = data$size,
      fill   = data$fill,
      alpha  = data$alpha,
      stroke = data$stroke
    )
    
    coords <- coord$transform(point_df, panel_params)
    ggplot2:::ggname(
      "geom_half_point",
      pointsGrob(
        coords$x, coords$y,
        pch = coords$shape,
        gp = gpar(
          col = alpha(coords$colour, coords$alpha),
          fill = alpha(coords$fill, coords$alpha),
          # Stroke is added around the outside of the point
          fontsize = coords$size * .pt + coords$stroke * .stroke / 2,
          lwd = coords$stroke * .stroke / 2
          )
        )
    )
  }
)
